core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -532,6 +563,33 @@ protected Type visitDereferenceExpression(DereferenceExpression node, StackableA
 
             // If this Dereference looks like column reference, try match it to column first.
             if (qualifiedName != null) {
+                // In the context of row pattern matching, fields are optionally prefixed with labels. Labels are irrelevant during type analysis.
+                if (context.getContext().isPatternRecognition()) {
+                    String label = label(qualifiedName.getOriginalParts().get(0));
+                    if (context.getContext().getLabels().contains(label)) {
+                        // In the context of row pattern matching, the name of row pattern input table cannot be used to qualify column names.
+                        // (it can only be accessed in PARTITION BY and ORDER BY clauses of MATCH_RECOGNIZE). Consequentially, if a dereference
+                        // expression starts with a label, the next part must be a column.
+                        // Only strict column references can be prefixed by label. Labeled references to row fields are not supported.
+                        QualifiedName unlabeledName = QualifiedName.of(qualifiedName.getOriginalParts().subList(1, qualifiedName.getOriginalParts().size()));
+                        if (qualifiedName.getOriginalParts().size() > 2) {
+                            throw semanticException(COLUMN_NOT_FOUND, node, "Column %s prefixed with label %s cannot be resolved", unlabeledName, label);
+                        }
+                        Expression unlabeled = DereferenceExpression.from(unlabeledName);
+                        Optional<ResolvedField> resolvedField = context.getContext().getScope().tryResolveField(unlabeled, unlabeledName);
+                        if (resolvedField.isEmpty()) {
+                            throw semanticException(COLUMN_NOT_FOUND, node, "Column %s prefixed with label %s cannot be resolved", unlabeledName, label);
+                        }
+                        // Correlation is not allowed in pattern recognition context. CorrelationSupport.DISALLOWED was set when creating the ExpressionAnalyzer,
+                        // and so the following call should fail if the field is from outer scope.
+                        Type type = process(unlabeled, new StackableAstVisitorContext<>(context.getContext().notExpectingLabels()));
+                        labelDereferences.put(NodeRef.of(node), new LabelPrefixedReference(label, unlabeled));
+                        return setExpressionType(node, type);
```

Comment:
Currently, we don't record relation type as ROW. Specifically, when we analyze a `DereferenceExpression`, we don't strip the part of the expression which is related to a relation. We could refactor it that way, but it is out of scope of MATCH_RECOGNIZE.

To avoid synthesizing an artificial expression, I used the original `Identifier` to analyze the label-prefixed column.